### DISCUSSION TITLE
Would it make sense to sign metadata on the original server to provided authentication?

### DISCUSSION THREAD

**Discussion by gdonval** [fwupd/fwupd#6151] [Q&A]
If I understood correctly, the current way things would work with `passim` is that authentication is provided by your TLS CA. I.e. you'll retrieve a hash using HTTPS. Without up-to-date `passim`-enabled peers, `fwupd` blindly trusts what is being downloaded over HTTPS.

In a word where some companies and "security" programs inject their own CA in the user's system I would feel much more at ease if fwupd did a signature check with public keys that would be shipped with the binary so that it provides its own root of trust.

From a more general perspective, this could allow you to lower the burden on a single server and allow people to provide metadata mirrors as an alternative to mDNS. It would still be over HTTPS, you *might* still want to retrieve a hash from the original server as an extra layer of security in the way you plan to do with `passim` but at least bandwidth burden could be shared.

**Comment by hughsie** [comment]
> authentication is provided by your TLS CA.

No, the certificate is self-signed and generated on first use.

> fwupd blindly trusts what is being downloaded over HTTPS

Well, I'm not sure I'd call it _blindly trusting_, but that's essentially true. The checksum is the minimum thing we use as trust.

> allow people to provide metadata mirrors as an alternative to mDNS

That's not what Passim tries to be; it's essentially LAN-only by design.

**Reply by gdonval** [reply]
> No, the certificate is self-signed and generated on first use.

There is a misunderstanding (and it may be on my side). I'm not talking about peer-to-peer communication.

In the current scheme, we get metadata from your CDN. We know it is the right data because of HTTPS.
In the new scheme, a peer that initiates a metadata update gets a checksum from the CDN over HTTPS (<- talking about this) then try to get the file locally through self-signed HTTPS (<- **not** talking about this), falling back to the CDN.

In that context, what tells `fwupd` "a file with this checksum is OK" (or "this file is OK" in the old scheme) is your CDN's TLS certificate.

If `fwupd` (as a binary file with local config) was that source of truth through public-encryption signatures you could think of delegating metadata delivery more easily, potentially without speaking to the CDN at all.

Does that make sense?

**Reply by hughsie** [reply]
> We know it is the right data because of HTTPS.

Ohh yes, I see. We also know it's the right data because it has to be signed by the GPG/PKCS#7 certs in `/etc/pki/fwupd-metadata/`

**Reply by gdonval** [reply]
Oh! So in principle, that signature and acquiring the hash are "redundant" for the purpose of identifying valid, untampered with, data, aren't they? If so that's all good! :)

**Reply by hughsie** [reply]
> So in principle, that signature and acquiring the hash are "redundant" for the purpose of identifying valid, untampered with, data, aren't they

I think so -- the TLS is more for encryption to avoid snooping -- and the checksum is redundant with the PKCS#7 signature -- but I didn't want to enforce that "all things added to passim must be signed" -- and it's also much more lightweight to do a simple sha256 before then checking the signature. Security in layers and all that.
